<!DOCTYPE html>




   <html class="no-js">

  <head lang="en-us">
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,user-scalable=no,initial-scale=1,maximum-scale=1">
    <meta http-equiv="X-UA-Compatible" content="IE=10" />
    <title>Transgenerational Inheritance - Inheritance of DNA Methylation</title>
    <meta name="generator" content="Hugo 0.21" />


    <meta name="description" content="Epigenotyping">

    <link rel="canonical" href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/transgenerational/">

    <meta name="author" content="Brigitte Hofmeister">


    <meta property="og:url" content="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/transgenerational/">
    <meta property="og:title" content="Inheritance of DNA Methylation">

    <meta name="apple-mobile-web-app-title" content="Inheritance of DNA Methylation">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">

    <link rel="shortcut icon" type="image/x-icon" href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/images/favicon.ico">
    <link rel="icon" type="image/x-icon" href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/images/favicon.ico">

    <style>
      @font-face {
        font-family: 'Icon';
        src: url('https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/fonts/icon.eot');
        src: url('https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/fonts/icon.eot')
               format('embedded-opentype'),
             url('https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/fonts/icon.woff')
               format('woff'),
             url('https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/fonts/icon.ttf')
               format('truetype'),
             url('https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/fonts/icon.svg')
               format('svg');
        font-weight: normal;
        font-style: normal;
      }
    </style>

    <link rel="stylesheet" href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/stylesheets/application.css">
    <link rel="stylesheet" href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/stylesheets/temporary.css">
    <link rel="stylesheet" href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/stylesheets/palettes.css">
    <link rel="stylesheet" href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/stylesheets/highlight/highlight.css">




    <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Roboto:400,700|Roboto&#43;Mono">
    <style>
      body, input {
        font-family: 'Roboto', Helvetica, Arial, sans-serif;
      }
      pre, code {
        font-family: 'Roboto Mono', 'Courier New', 'Courier', monospace;
      }
    </style>


    <link rel="stylesheet" href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/custom.css">

    <script src="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/javascripts/modernizr.js"></script>



  </head>
  <body class="palette-primary-cyan palette-accent-red">







<div class="backdrop">
	<div class="backdrop-paper"></div>
</div>

<input class="toggle" type="checkbox" id="toggle-drawer">
<input class="toggle" type="checkbox" id="toggle-search">
<label class="toggle-button overlay" for="toggle-drawer"></label>

<header class="header">
	<nav aria-label="Header">
  <div class="bar default">
    <div class="button button-menu" role="button" aria-label="Menu">
      <label class="toggle-button icon icon-menu" for="toggle-drawer">
        <span></span>
      </label>
    </div>
    <div class="stretch">
      <div class="title">
        Transgenerational Inheritance
      </div>
    </div>






  </div>
  <div class="bar search">
    <div class="button button-close" role="button" aria-label="Close">
      <label class="toggle-button icon icon-back" for="toggle-search"></label>
    </div>
    <div class="stretch">
      <div class="field">
        <input class="query" type="text" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck>
      </div>
    </div>
    <div class="button button-reset" role="button" aria-label="Search">
      <button class="toggle-button icon icon-close" id="reset-search"></button>
    </div>
  </div>
</nav>
</header>

<main class="main">
	<div class="drawer">
		<nav aria-label="Navigation">
  <a href="https://github.com/schmitzlab/stable-transgenerational-inheritance-of-DNA-methylation" class="project">
    <div class="banner">

      <div class="name">
        <strong>Inheritance of DNA Methylation </strong>

          <br>
          schmitzlab/stable-transgenerational-inheritance-of-DNA-methylation

      </div>
    </div>
  </a>

  <div class="scrollable">
    <div class="wrapper">

        <ul class="repo">
          <li class="repo-download">
            <a href="https://github.com/schmitzlab/stable-transgenerational-inheritance-of-DNA-methylation/archive/master.zip" target="_blank" title="Download" data-action="download">
              <i class="icon icon-download"></i> Download
            </a>
          </li>

            <li class="repo-stars">
            <a href="https://github.com/schmitzlab/stable-transgenerational-inheritance-of-DNA-methylation/stargazers" target="_blank" title="Stargazers" data-action="star">
              <i class="icon icon-star"></i> Stars
              <span class="count">&ndash;</span>
            </a>
          </li>
        </ul>
        <hr>


      <div class="toc">

        <ul>





<li>





<a  title="Home" href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/">

	Home
</a>




</li>



<li>





<a  title="Getting started" href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/getting-started/">

	Getting started
</a>




</li>



<li>





<a class="current" title="Transgeneration inheritance" href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/transgenerational/">

	Transgeneration inheritance
</a>


<ul id="scrollspy">
</ul>



</li>



<li>





<a  title="Epigenotyping" href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/epigenotyping/">

	Epigenotyping
</a>




</li>



<li>





<a  title="Appendix" href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/appendix">

	Appendix
</a>




</li>


        </ul>



        <hr>
        <span class="section">The author</span>

        <ul>





        </ul>

      </div>
    </div>
  </div>
</nav>

	</div>

	<article class="article">
		<div class="wrapper">
			<h1>Transgenerational Inheritance </h1>



<p><strong>Sample data is not provided for this topic.</strong>
For illustrative purposes, assume that three generations of line &ldquo;A&rdquo; are available and they are named &ldquo;A-1&rdquo;, &ldquo;A-2&rdquo;, &ldquo;A-3&rdquo;. This species only has two chromosomes, &ldquo;Chr1&rdquo; and &ldquo;Chr2&rdquo;.</p>

<h2 id="beginnings">Beginnings</h2>

<h3 id="set-up-requirements">Set up requirements</h3>

<ul>
<li>All samples have allC files.

<ul>
<li>AllC file are tab-delimited.</li>
<li>They have the following columns

<ul>
<li>Chromosome</li>
<li>Position (1-indexed)</li>
<li>Strand (+ or -)</li>
<li>Context (3 bp sequence)</li>
<li>Number methylated reads</li>
<li>Number total reads</li>
<li>Is the position methylated? (1 for yes)</li>
</ul></li>
</ul></li>
<li>All of the allC files are named as &ldquo;allc_samplename_chromosome.tsv&rdquo; or &ldquo;allc_samplename.tsv&rdquo;.</li>
<li>Many of the following scripts expect a single allC file per sample. In this case, the file is named &ldquo;allc_samplename.tsv&rdquo;. Create this file using <code>cat allc_samplename_*.tsv &gt; allc_samplename.tsv</code>.</li>
<li>All of the allC files are in one directory.</li>
</ul>

<h2 id="identification-of-methylated-regions">Identification of Methylated Regions</h2>

<h3 id="step-1-combine-samples-allc-files">Step 1: Combine samples allC files</h3>

<p>We combine the information from all samples in a line to form a &ldquo;pan-methylome&rdquo;. To do this, we sum up the number of methylated and number of total reads at each position across all samples and write the results to a new allC file.</p>

<p>For example run, <a href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/appendix/#combine-allc-pe-py">combine_allc_pe.py</a></p>

<pre><code class="language-bash">python combine_allc_pe.py -o=lineA-pan-methylome -c=Chr1,Chr2 allc_path A-1 A-2 A-3
</code></pre>

<p>This creates two files (one for each chromosome) <em>allc_lineA-pan-methylome_Chr1.tsv</em> and <em>allc_lineA-pan-methylome_Chr2.tsv</em>.</p>

<h3 id="step-2-unmethylate-the-pan-methylome">Step 2: Unmethylate the pan-methylome</h3>

<p>We are trying to identify all methylated regions, so we want to create a null, unmethylated genome to compare it against.</p>

<pre><code class="language-bash">python unmethylate_allc_pe.py allc_lineA-pan-methylome_Chr1.tsv allc_lineA-pan-methylome_Chr2.tsv
</code></pre>

<p>This creates two files (one for each input): <em>allc_lineA-pan-methylome-unmethylated_Chr1.tsv</em> and <em>allc_lineA-pan-methylome-unmethylated_Chr2.tsv</em></p>

<h3 id="step-3-dmr-analysis">Step 3: DMR analysis</h3>

<p>Using your prefer DMR analysis tool, you can compare the pan-methylome and unmethylated pan-methylome to identify methylated regions of the genome. Parameters and filters should be the same for this analysis as used for between-samples DMR analysis.</p>

<h2 id="epiallele-identification">Epiallele Identification</h2>

<h3 id="step-1-putative-dmr-list">Step 1: Putative DMR list</h3>

<p>Using the preferred DMR analysis tool, identify DMRs between all samples in a line. Depending on the program, this DMR list should be filtered for length, methylation change, ect.
After filtering, we have a set of putative epialleles. For illustrative purposes, these regions are saved in a file named <em>dmr_list.tsv</em>.</p>

<h3 id="step-2-coverage-filters">Step 2: Coverage filters</h3>

<p>Many of the scripts in this analysis accept an optional minimum coverage parameter which only keeps positions with sufficient coverage in all samples. We apply this minimum coverage to minimize methylation variation of a region/sample due to missing information.</p>

<p>To save computational time later, we will create a new allC file for each sample which only includes positions with sufficient coverage. This eliminates needing to apply this filter repetitively.</p>

<p>For example, run <a href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/appendix/#filter-allc-coverage-pe-py">filter_allc_coverage_pe.py</a> to with minimum coverage 2. This script expects all chromosomes in one allC file.</p>

<pre><code class="language-bash">python filter_allc_coverage_pe.py -v=2 allc_path A-1 A-2 A-3
</code></pre>

<p>This creates files <em>allc_A-1_cov2.tsv</em>, <em>allc_A-2_cov2.tsv</em>, and <em>allc_A-3_cov2.tsv</em>.</p>

<h3 id="step-3-between-generation-comparison">Step 3: Between generation comparison</h3>

<p>With the DMR list, we want to get the number of methylated and unmethylated reads for each region and compare adjacent generations. First, we will get the read counts for organized by each comparison. When specifying samples, <strong>order matters</strong>.</p>

<p>To allow for fair comparsion between samples, we want to use positions which have sufficient coverage in all samples. So we use the coverage allC files created in Step 2. In this script, the coverage parameter is used to search for files named as <code>allc_samplename_cov#.tsv</code>, which includes all chromosomes.</p>

<p>Run <a href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/appendix/#dmr-gen-counts-pe-py">dmr_gen_counts_pe.py</a> for minimum coverage 2.</p>

<pre><code class="language-bash">python dmr_gen_counts_pe.py -v=2 dmr_list.tsv allc_path A-1 A-2 A-3
</code></pre>

<p>This creates file <em>out_dmr-gen_c.tsv</em>.</p>

<h3 id="step-4-identify-signficant-dmrs">Step 4. Identify signficant DMRs</h3>

<p>Next, we use a z-test to determine if a a region is differentially methylated between two generations. We are testing for a change in methylation level greater a specified threshold.
The threshold is important because we could identify a region with statistically significant differential methylation but is only a change of 1%, which is unlikely to be biologically signficant. Threshold is definited as
* Percent change between samples, i.e. 40% to 50% is a 25% change (default)
* Raw methylation level, i.e. 40% to 50% is a 10% change
Also includes parameters to set minimum length of region and minimum number of covered cytosines.</p>

<p>This script uses the output of Step 3, <em>out_dmr-gen_c.tsv</em>.</p>

<p>Run <a href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/appendix/#dmr-gen-ztesting-py">dmr_gen_ztesting.py</a> to test for methylation level change of 25%</p>

<pre><code class="language-bash">python dmr_gen_ztesting.py -wm -m=0.25 out_dmr-gen_c.tsv
</code></pre>

<p>This creates three files: <em>out_dmr-gen_c_full.tsv</em> (full set of statistics for input dataset), <em>out_dmr-gen_c_switches.tsv</em> (number of switches per region), and <em>out_dmr-gen_c_switch_counts.tsv</em> (number of regions per number of possible switches).</p>

<h3 id="step-5-convert-to-bed-file">Step 5: Convert to BED file</h3>

<p>Now, we use the output of Step 4 (<em>out_dmr-gen_C_switches.tsv</em>) to create a BED file of the DMRs. Score parameter filters based on the number of switches.</p>

<p>Run <a href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/appendix/#dmr-file-to-bed-py">dmr_file_to_bed.py</a> to get all regions with at least one switch</p>

<pre><code class="language-bash">python dmr_file_to_bed.py -v=1 out_dmr-gen_c_switches.tsv
</code></pre>

<p>This creates <em>out_dmr-gen_c.bed</em>.</p>


			<aside class="copyright" role="note">

				&copy; 2017 Released under the MIT license &ndash;

				Documentation built with
				<a href="https://www.gohugo.io" target="_blank">Hugo</a>
				using the
				<a href="http://github.com/digitalcraftsman/hugo-material-docs" target="_blank">Material</a> theme.
			</aside>

			<footer class="footer">


<nav class="pagination" aria-label="Footer">
  <div class="previous">

      <a href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/getting-started/" title="Getting started">
        <span class="direction">
          Previous
        </span>
        <div class="page">
          <div class="button button-previous" role="button" aria-label="Previous">
            <i class="icon icon-back"></i>
          </div>
          <div class="stretch">
            <div class="title">
              Getting started
            </div>
          </div>
        </div>
      </a>

  </div>

  <div class="next">

      <a href="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/epigenotyping/" title="Epigenotyping">
        <span class="direction">
          Next
        </span>
        <div class="page">
          <div class="stretch">
            <div class="title">
              Epigenotyping
            </div>
          </div>
          <div class="button button-next" role="button" aria-label="Next">
            <i class="icon icon-forward"></i>
          </div>
        </div>
      </a>

  </div>
</nav>





			</footer>
		</div>
	</article>

	<div class="results" role="status" aria-live="polite">
		<div class="scrollable">
			<div class="wrapper">
				<div class="meta"></div>
				<div class="list"></div>
			</div>
		</div>
	</div>
</main>

    <script>

      var base_url = 'https:\/\/schmitzlab.github.io\/stable-transgenerational-inheritance-of-DNA-methylation\/';
      var repo_id  = 'schmitzlab\/stable-transgenerational-inheritance-of-DNA-methylation';

    </script>

    <script src="https://schmitzlab.github.io/stable-transgenerational-inheritance-of-DNA-methylation/javascripts/application.js"></script>


    <script>
      /* Add headers to scrollspy */
      var headers   = document.getElementsByTagName("h2");
      var scrollspy = document.getElementById('scrollspy');

      if(scrollspy) {
        if(headers.length > 0) {
          for(var i = 0; i < headers.length; i++) {
            var li = document.createElement("li");
            li.setAttribute("class", "anchor");

            var a  = document.createElement("a");
            a.setAttribute("href", "#" + headers[i].id);
            a.setAttribute("title", headers[i].innerHTML);
            a.innerHTML = headers[i].innerHTML;

            li.appendChild(a)
            scrollspy.appendChild(li);
          }
        } else {
          scrollspy.parentElement.removeChild(scrollspy)
        }


        /* Add permanent link next to the headers */
        var headers = document.querySelectorAll("h1, h2, h3, h4, h5, h6");

        for(var i = 0; i < headers.length; i++) {
            var a = document.createElement("a");
            a.setAttribute("class", "headerlink");
            a.setAttribute("href", "#" + headers[i].id);
            a.setAttribute("title", "Permanent link")
            a.innerHTML = "#";
            headers[i].appendChild(a);
        }
      }
    </script>



    <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.8.0/highlight.min.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
  </body>
</html>

